### ISSUE TITLE
Enhanced analysis of variadic references

### ISSUE THREAD

**Issue by bugreportuser** [vimeo/psalm#1473]
There are some problems when a parameter is variadic and takes references:
1. The PHPDoc param is parsed inconsistently.
   - `type ...&$x` - marked wrong and ignored
   - `type &...$x` - ignored (I'm not saying this should be supported, but it's worth mentioning.)
   - `type ...$x` - no problem
   - `type[] &$x` - no problem
2. The variables used as arguments become arrays outside the function. I'm guessing this is because the parameter type is changed to `array` to analyze the body of the function.

Example (https://psalm.dev/r/9f9eecf235):
```php
<?php
/**
 * @param int ...&$x
 */
function example (&...$x): void {}

$x = 0;
example($x);
chr($x);
$x = 0;
```

Example output:
```
Psalm output (using commit ab83220):

INFO: InvalidDocblockParamName - 3:19 - Incorrect param name $$x in docblock for example

ERROR: InvalidArgument - 9:5 - Argument 1 of chr expects int, array<int, mixed> provided

ERROR: ReferenceConstraintViolation - 10:1 - Variable $x is limited to values of type array<int, mixed> because it is passed by reference, int(0) type found
```

**Comment by muglug** [comment]
I've addressed the above point 2 in 1348d63, so a regular `int` typehint works as expected:

```php
<?php
function example (int &...$x): void {}

$x = 0;
example($x);
chr($x);
$x = 0;
```
https://psalm.dev/r/40b9824928

**Comment by bugreportuser** [comment]
Thanks!
